src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp

```diff
@@ -1745,7 +1757,85 @@ void CompositionalMultiphaseBase::applyDirichletBC( real64 const time_n,
         } );
       } );
     }
+  } );
+}
+
+void CompositionalMultiphaseBase::keepFlowVariablesConstantDuringStep( real64 const time,
+                                                                       real64 const dt,
+                                                                       DofManager const & dofManager,
+                                                                       DomainPartition & domain,
+                                                                       CRSMatrixView< real64, globalIndex const > const & localMatrix,
+                                                                       arrayView1d< real64 > const & localRhs ) const
+{
+  GEOSX_MARK_FUNCTION;
+
+  GEOSX_UNUSED_VAR( time, dt );
+
+  forDiscretizationOnMeshTargets( domain.getMeshBodies(), [&]( string const &,
+                                                               MeshLevel const & mesh,
+                                                               arrayView1d< string const > const & regionNames )
+  {
+    mesh.getElemManager().forElementSubRegions( regionNames,
+                                                [&]( localIndex const,
+                                                     ElementSubRegionBase const & subRegion )
+    {
+      globalIndex const rankOffset = dofManager.rankOffset();
+      string const dofKey = dofManager.getKey( viewKeyStruct::elemDofFieldString() );
 
+      arrayView1d< integer const > const ghostRank = subRegion.ghostRank();
+      arrayView1d< globalIndex const > const dofNumber = subRegion.getReference< array1d< globalIndex > >( dofKey );
+
+      arrayView1d< real64 const > const pres = subRegion.getField< fields::flow::pressure >();
+      arrayView1d< real64 const > const temp = subRegion.getField< fields::flow::temperature >();
+      arrayView2d< real64 const, compflow::USD_COMP > const compDens = subRegion.getField< fields::flow::globalCompDensity >();
+
+      integer const numComp = m_numComponents;
+      integer const isThermal = m_isThermal;
+      forAll< parallelDevicePolicy<> >( subRegion.size(), [=] GEOSX_HOST_DEVICE ( localIndex const ei )
+      {
+        if( ghostRank[ei] >= 0 )
+        {
+          return;
+        }
+
+        globalIndex const dofIndex = dofNumber[ei];
+        localIndex const localRow = dofIndex - rankOffset;
```

Comment:
Maybe it's not really for this PR, but could we get this pattern embedded/unified somewhere.
Maybe like `localIndex const localRow = dofManager.buildLocalRow( dofIndex )`?
So we could document what going on. But maybe there are some other constraints also...

Reply:
Yes, we could do something like that, but we do not want to bring the `DofManager` in the kernel. 

So either we would compute a vector of `localRow`s in advance, or the `DofManager` would have some kind of kernel wrapper to do `localRow = dofIndex - rankOffset`. It may be an overkill, but maybe there is another way to unify this that I don't see.

Reply:
I guess it's OK like this.